ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/DeduplicationState.scala

```diff
@@ -51,15 +53,80 @@ case class DeduplicationState private (
 }
 
 object DeduplicationState {
-  private[sandbox] type DeduplicationQueue = VectorMap[ChangeId, Time.Timestamp]
+  private[sandbox] type DeduplicationQueue = UnsafeDeduplicationStateQueueMap
 
   private[validate] def empty(
       deduplicationDuration: Duration,
       bridgeMetrics: BridgeMetrics,
   ): DeduplicationState =
     DeduplicationState(
-      deduplicationQueue = VectorMap.empty,
+      deduplicationQueue = UnsafeDeduplicationStateQueueMap.empty,
       maxDeduplicationDuration = deduplicationDuration,
       bridgeMetrics = bridgeMetrics,
     )
+
+  /** This data structure is tailored for keeping an ordered (by insertion) sequence of deduplication entries
+    * of the form (changeId, recordTime) with optimal complexity of each exposed operation.
+    *
+    * @param vector An ordered (by insertion) vector of deduplication entries
+    * @param mappings Mapping of changeId to recordTime
+    */
+  private[validate] case class UnsafeDeduplicationStateQueueMap(
+      vector: Vector[(ChangeId, Time.Timestamp)],
+      mappings: Map[ChangeId, Time.Timestamp],
+  ) {
+
+    /** Returns a state without the entries that are before the expirationTimestamp.
+      *
+      * Complexity: eL - effectively linear in the number of expired entries
+      */
+    def withoutOlderThan(expirationTimestamp: Time.Timestamp): UnsafeDeduplicationStateQueueMap = {
+      // Assuming that the entries are monotonically increasing with regard to the recordTime,
+      // get all entries with the recordTime before the expirationTimestamp
+      val expiredFromVector = vector.view.takeWhile { case (_, recordTime) =>
+        recordTime < expirationTimestamp
+      }
+
+      // A recordTime for a changeId could have been updated by a newer command inserted in the mapping.
+      // Hence, remove only entries whose recordTimes match the expired entry's recordTime
+      val expiredFromMappings = expiredFromVector.flatMap { case (changeId, recordTime) =>
+        mappings.get(changeId).filter(_ == recordTime).map(_ => changeId)
+      }
+
+      UnsafeDeduplicationStateQueueMap(
+        vector = vector.drop(expiredFromVector.size),
```

Comment:
not sure this was intentional or not but since you are using a view, this .size will traverse the vector the second time

Reply:
You're right, but am not sure how to avoid it since early materialization of the vector would add the burden somewhere else. If only Scala would have a `Vector#splitAtCondition` (similar to `splitAt` but vector based)

Reply:
you could start by a dropWhile instead of takeWhile, and materialize the final new version of the vector at the beginning
if you do that you have the count of the new vector as well
and still the old vector:
which you can .view.take(vector.size - newVector.size) which would give you the lazy diff vector for map houskeeping

Reply:
Ok, used as such e0a354e086cdcfda2e09f0791f1d2babf9bc43e3